### ISSUE TITLE
Error when bind a QObject inherited interface to a class that has a dependency

### ISSUE THREAD

**Issue by LempereurBenjamin** [boost-ext/di#333]
Here is the diagram of the situation :
![test](https://user-images.githubusercontent.com/6420842/39628016-f99a6a7a-4fa7-11e8-92d7-ad2dfd5e76ce.png)

Here is the error message :
```
In file included from /home/blm/poc-boost-di/main.cpp:4:0:
/home/blm/poc-boost-di/./QObject/Example.h: In function ‘void QtPoc::Example()’:
/home/blm/poc-boost-di/./QObject/Example.h:19:37: error: ‘T boost::di::v1_0_1::core::injector<TConfig, boost::di::v1_0_1::core::pool<boost::di::v1_0_1::aux::type_list<> >, TDeps ...>::create() const [with T = QtPoc::B; typename boost::di::v1_0_1::aux::enable_if<(! boost::di::v1_0_1::core::injector<TConfig, boost::di::v1_0_1::core::pool<boost::di::v1_0_1::aux::type_list<> >, TDeps ...>::is_creatable<T, boost::di::v1_0_1::no_name, boost::di::v1_0_1::aux::integral_constant<bool, true> >::value), int>::type <anonymous> = 0; TConfig = boost::di::v1_0_1::config; TDeps = {boost::di::v1_0_1::core::dependency<boost::di::v1_0_1::scopes::singleton, QtPoc::IA, QtPoc::A, boost::di::v1_0_1::no_name, void>, boost::di::v1_0_1::core::dependency<boost::di::v1_0_1::scopes::singleton, QtPoc::IC, QtPoc::C, boost::di::v1_0_1::no_name, void>}]’ is deprecated: creatable constraint not satisfied [-Werror=deprecated-declarations]
          auto b = injector.create<B>();
                                    ^
 In file included from /home/blm/poc-boost-di/./QObject/Example.h:5:0,
                  from /home/blm/poc-boost-di/main.cpp:4:
/home/blm/poc-boost-di/./boost/di.hpp:2417:3: note: declared here
create
 ^
/home/blm/poc-boost-di/./boost/di.hpp: At global scope:
/home/blm/poc-boost-di/./boost/di.hpp:891:2: error: inline function ‘static T* boost::di::v1_0_1::concepts::abstract_type<T>::is_not_bound::error(boost::di::v1_0_1::_) [with T = QtPoc::A]’ used but never defined [-Werror]
error(_ = "type is not bound, did you forget to add: 'di::bind<interface>.to<implementation>()'?");
^
cc1plus: all warnings being treated as errors
```
But when I remove the IC& dependency in A or remove the QObject inheritance from IA, I don't have any error.
![test2](https://user-images.githubusercontent.com/6420842/39628122-4e6e599e-4fa8-11e8-9b25-92dc547009df.png)

Injection code : 
```
        auto injector = boost::di::make_injector
                (
                        boost::di::bind<IA>().to<A>().in(boost::di::singleton),
                        boost::di::bind<IC>().to<C>().in(boost::di::singleton)
                );
        auto b = injector.create<B>();

```

How can I avoid this error ?

**Comment by kris-jusiak** [comment]
Thanks @LempereurBenjamin.

Would you be able to provide code examples, please?
I really like your diagrams but it takes a while to figure out the problems.

I think that you need both IA and A to be resolved. If that's the case you would have to allow both to be resolved like that:
```cpp
boost::di::bind<IA, A>.to<A>(); // this way IA and A will be able to be resolved
```

**Comment by LempereurBenjamin** [comment]
@krzysztof-jusiak I tried to resolve A class too but already the same error.
Here is the example code : 
```
 #pragma once

#include <QtCore/QObject>

namespace QtPoc
{
    class IA : public QObject
    {
        Q_OBJECT
    public:
        virtual ~IA() = default;
        virtual void Child() = 0;

        signals:
        void Signal();
    };
}
```

```
#pragma once

#include "IC.h"
#include "IA.h"

namespace QtPoc {
class A : public IA
{
public:
    A(IC& c);
    void Child() override
    {}

private:
    IC& _c;
};}
```

```
#pragma once

#include "IA.h"
#include "IC.h"

namespace QtPoc {
    class C : public IC
    {
    };}
```

```
namespace QtPoc
{
    void Example()
    {
        auto injector = boost::di::make_injector
                (
                        boost::di::bind<IA, A>().to<A>().in(boost::di::singleton),
                        boost::di::bind<IC>().to<C>().in(boost::di::singleton)
                );
        auto b = injector.create<B>();
    }
}
```

Thank you !

**Comment by LempereurBenjamin** [comment]
@krzysztof-jusiak If I add a default constructor and I replace the C& with a shared_ptr<C> in A class, it works ! Do you have any idea why ?

**Comment by kris-jusiak** [comment]
@LempereurBenjamin, Can you help me with reproducing the issue, please?

I godbolt your example and seem to work, so I'm assuming your case is a bit different:
https://godbolt.org/g/7jiwsK

Thanks.

**Comment by LempereurBenjamin** [comment]
The issue append only when the "IA" interface inherit from QObject. But QT is not available in Godbolt ?
I fixed the issue by using the BOOST_DI_INJECT macro in the "A" class to force the usage of the specified constructor but i don't know why it's needed.

```
#pragma once

#include "IC.h"
#include "IA.h"

namespace QtPoc {
class A : public IA
{
public:
   BOOST_DI_INJECT(A, IC& c) : _c(c) {}
    void Child() override
    {}

private:
    IC& _c;
};
}
```

**Comment by kris-jusiak** [comment]
Hmm,

The following code compiles for me:
```cpp
#include <boost/di.hpp>
#include <Qt/qobject.h>
#include <memory>

namespace QtPoc
{
  struct QObjectPrivate{};
    class IA : public QObject
    {
       //Q_OBJECT
    public:
        virtual ~IA() = default;
        virtual void Child() = 0;

  //      signals:
        //void Signal();
    };
    class IC : public QObject
    {
       //Q_OBJECT
    public:
        virtual ~IC() = default;
        virtual void Child() = 0;
  //      signals:
        //void Signal();
    };
}

namespace QtPoc {
class A : public IA
{
public:
    A(IC& c) : _c{c} {}

    void Child() override
    {}

private:
    IC& _c;
};
}

namespace QtPoc {
    class C : public IC
    {
      void Child() override
      {}
    };
}

namespace QtPoc {
struct B {
  B(IA& a) : a{a}{}
  IA& a;
};
}

namespace QtPoc
{
    void Example()
    {
        auto injector = boost::di::make_injector
                (
                        boost::di::bind<IA>().to<A>().in(boost::di::singleton),
                        boost::di::bind<IC>().to<C>().in(boost::di::singleton)
                );
        auto b = injector.create<B>();
    }
}
int main() {
  QtPoc::Example();
}
```

Can you double check whether including a <memory> header changes the behaviour, please?

**Comment by LempereurBenjamin** [comment]
I'm already facing this issue. Is a compiler issue possible ? I'm using GCC 4.8.5

**Comment by kris-jusiak** [comment]
Possibly. The library doesn't support GCC < 5 officially, hence it's untested ground. I'm not even sure that 4.8.5 supports C++14?
I tried to compile the following guy but `__cplusplus` is not set for C++14 (even with C++1y) -> https://godbolt.org/g/UJBwxo

**Comment by LempereurBenjamin** [comment]
I tried to upgrade to 5.3.1 but i had already the same issue. Is the Qt version can be the problem ?

**Comment by kris-jusiak** [comment]
Woohoo, I reproduced the issue with DI version 1.0.1 (It happens with Qt4 and Qt5).
The issue doesn't seem to happen in the version 1.0.2 and/or the newest version available in the repository.
Can you please try the newest release 1.0.2 or the newest version of the repo, please?

**Comment by LempereurBenjamin** [comment]
Perfect, It's fixed in in the version 1.0.2. Thank you for your help.
